_posts/2018-10-15-12-2-Use-SkipToken-for-Server-Side-Paging.md

```diff
@@ -0,0 +1,94 @@
+
+#### Generating the nextlink
+The next link generation method in ___GetNextPageHelper___ static class will take in the $skiptoken value along with other query parameters and generate the link by doing special handling for $skip, $skiptoken and $top. It will pass on the other query options as they were in the original request.
+
+##### 1. Handle $skip
+We will omit the $skip value if the service is configured to support $skiptoken and a collection of entity is being requested. This is because the first response would have applied the $skip query option to the results already. 
+##### 2. Handle $top
+We will reduce the value of $top query option by the page size if it is greater than the page size.   
+##### 3. Handle $skiptoken
+The value for the $skiptoken will be updated to new value passed in which is the key value for the last record sent. If the skiptoken value is not sent, we will call the existing method and use $skip for paging instead.
+
+#### Routing
+Since we will only be modifying the query options from the original request to generate the nextlink, the routing will remain same as the original request. 
+
+#### Parsing $skiptoken and generating the Linq expression
+New classes will be created for ___SkipTokenQueryOption___ and __SkipTokenQueryValidator__. ___SkipTokenQueryOption___ will contain the  methods to create and apply the LINQ expression based on the $skiptoken value. To give an example, for a query like the following:
+
+`GET ~/EntitySet?$orderby=Prop1,Prop2&$skiptoken=Prop1:value1,Prop2:value2,Id1:idVal1,Id2:idVal2`
+
+The following where clause will be added to the predicate:
+```
+WHERE Prop1>value1
+Or (Prop1=value1 AND Prop2>value2)
+Or (Prop1=value1 AND Prop2=value2 AND Id1>Val)
+Or (Prop1=value1 AND Prop2=value2 AND Id1=idVal1 AND Id2>idVal2)
+```
+Note that the greater than operator will be swapped for less than operator if the order is descending. 
+#### Generating the $skiptoken
+The ___SkipTokenQueryOption___ class will be utilized by ___ODataQueryOption___ to pass the token value to the nextlink generator helper methods.
+In the process, ___IWebApiRequestMessage___ will be modified and GetNextPageLink method will be overloaded to now accept another parameter for the $skiptoken value.
+
+#### Configuration to use $skiptoken or $skip for server-driven paging
+We will allow services to configure if they want to use $skiptoken or $skip for paging per route as there can be performance issues with a large database with multipart keys. By default, we will use $skip.
+
+Moreover, we will ensure stable sorting if the query is configured for using $skiptoken. 
+### Additional details and discussions
+#### 1.	How would a developer implement paging without using EnableQuery attribute? What about stable ordering in that case?
+a.	 The new SkipTokenQueryOption class will provide 2 methods-
+
+      i.	GenerateSkipTokenValue – requires the EDM model, the results as IQuerable and OrderbyQueryOption.
```

Comment:
>GenerateSkipTokenValue [](start = 9, length = 22)

How do I parse the SkipTokenValue? Are there methods for extracting the property names/values? The LINQ expression?

Reply:
There were private methods needed but I believe exposing them will make it easier for developers.